judge/models/problem.py

```diff
@@ -407,6 +407,25 @@ def save(self, *args, **kwargs):
 
     save.alters_data = True
 
+    def can_vote(self, user):
+        if not user.is_authenticated:  # reject anons
+            return False
+
+        # if user is unlisted or is banned from submitting to the problem, then they cannot vote
+        if user.profile.is_unlisted or self.banned_users.filter(pk=user.pk).exists():
```

Comment:
I would say to place these ifs in the following order:
```
check user.profile.current_contest
check user.profile.is_unlisted
check user.profile.is_banned_from_voting_problem_points
check banned users
```
to save on a query checking `banned_users` since it's much more likely that a user is in a contest or unlisted (and therefore short circuits without querying the database) rather than banned from the problem.